### PR TITLE
 Decentralized adaptive optimizers

### LINKED ISSUE #184: Support communication-efficient optimizers
`<edit>`
The goal of this project is to implement pre-configured decentralized optimizers such as [DADAM](https://arxiv.org/abs/1901.09109) and decentralized [Local AMSGrad](https://dl.acm.org/doi/10.1145/3412815.3416891). We intend to create a general interface that runs adaptive optimizers with periodic averaging of gradients, parameters and optimizer statistics on different time-frames]

The final version should support:
* [ ] wrapping any adam-like optimizers such as Adam, AMSGrad, LAMB(FusedLAMB), Radam, etc
* [ ] full torch.optim.Optimizer interface including state_dict/load_state_dict
   * [ ] for parameter groups, either support it or explicitly throw an error
* [ ] averaging optimizer statistics in background while the model is training, incorporating local updates after the averaging is done
* [ ] have a default confguration that converges :)
  * [ ] test convergence locally with a toy task
  * [ ] test in standard ALBERT-base setup for wikitext103
  * [ ] test with ALBERT-large on OpenBookCorpus

`</edit>`

Code draft:

```
# usage:
class LAMBWithAveraging(FusedLAMB, AdaptiveOptimizer):
    def get_averaged_statistics() -> Sequence[Tensor]
       return TODO()

opt = DecentralizedOptimizer(model, base_optimizer=LAMBWithAveraging(model.parameters()),
                             prefix='MY_UNIQUE_STRING')


### code

class AdaptiveOptimizer(abc.Mixin):
   def get_averaged_statistics(self):
     raise NotImplementedError()

class DecentralizedOptimizer(abc.meta):
    def __init__(self, parameters, base_optimizer, dht, prefix: str, ..., ):
       assert isinstance(base_optimizer, AdaptiveOptimizer), "TODO see example here"

       self.parameters = list(parameters)  # on GPU
       self.base_optimizer = base_optimizer
       self.global_step = ...
       
       zeros_like_grads = [torch.zeros_like(x.detach().cpu()) for x in self.parameters]
       self.grad_averager = DecentralizedAverager(
           averaged_tensors= zeros_like_grads  # average gradients only
            prefix=prefix+'_grad_averager', **some_kwargs
           )
       
       parameters = [x.cpu().detach().float() self.parameters()]
       averaged_statistics = [x.cpu().detach().float() for x in base_optimizer.get_averaged_statistics()]
       
       self.almost_v_averager = DecentralizedAverager(
           averaged_tensors= parameters + averaged_statistics  # average params and second momenta
            prefix=prefix+'_v_averager', **some_other_kwargs
           )
       self.v_averaging_job = None
       
    def get_model_params_and_v(self) -> Tuple[torch.Tensor, ...]:
        TODO()
              
    def step(self):
       accumulate_grads()
       report_progress_to_dht()
       
       if self.v_averaging_job.done():
          local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
          with self.almost_v_averager.get_tensors() as averaged_params_and_v:
             copy_from(averaged_params_and_v).to(local_params_and_v)

       
       if accumulated_enough():
       
           local_grads = [x.grad for x in model.parameters()]
           with self.grad_averager.get_tensors() as averaged_grads:
               copy_from(local_grads).to(averaged_grads)
           self.grad_averager.step()
           with self.grad_averager.get_tensors() as averaged_grads:
               copy_from(averaged_grads).to(local_grads)
           # TODO maybe average gradients continuously in background
             
           if not self.use_local_statistics:
               TODO(revert base_optimizer.get_averaged_statistics() to prev averaged state)
           self.base_optimizer.step()
           self.global_step += 1  # MAKE SURE IT IS in sync with collaboration
           
           if not self.use_local_statistics:
               TODO(revert optimizer statistics back to local ones AFTER local update)
           
           
           ### AVERAGE PARAMETERS AND STATISTICS! ###
           if self.global_step % self.K == 0:
               if not v_averaging_job.done():
                   if self.await_v_averager:
                      v_averaging_job.wait()
                      local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
                      with self.almost_v_averager.get_tensors() as averaged_params_and_v:
                         copy_from(averaged_params_and_v).to(local_params_and_v)
            
                   else:
                       warning("Param averager is taking too long, TODO explain why we skip")
                       return
           
               local_params_and_v = self.paramters + self.base_optimzier.get_averaged_stats()
               with self.almost_v_averager.get_tensors() as averaged_params_and_v:
                   copy_from(local_params_and_v).to(averaged_params_and_v)
               
               self.v_averaging_job = run_in_background(almost_v_averager.step())
           
       
    def zero_grad(self):
       self.opt.zero_grad()
       zero_grads_in_self_grad_averager()
       self.report_to_dht()
```

### PR DESCRIPTION
**Pull request by nevec**
This PR adds support for decentralized adaptive optimizer and hopefully closes #184 
- transforms `ParameterAveragingOptimizer` into `DecentralizedOptimizer` which supports many averaging options 
- adds preconfigured DADAM
- adds `CollaborativeAdaptiveOptimizer` which accounts for adaptive learning rates

### CODE REVIEW

**Review by xtinkt — APPROVED** [review]
LGTM

File: hivemind/client/averaging/training.py [hunk 1]
```diff
@@ -46,27 +48,51 @@ def __init__(self, opt: torch.optim.Optimizer, *, average_parameters: bool, aver
...
-                    local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype, device=local_tensor.device)
+            if gathered is not None:
+                # load averaged tensors back into model
+                with data_lock, self.get_tensors() as averaged_tensors:
+                    assert len(averaged_tensors) == len(
+                        local_tensors), "The number of optimized parameters should not change."
+
+                    if use_old_local_tensors:
+                        # since tensors might have changed, we subtract old_local_tensor and add averaged. This prevents
+                        # from loosing local updates that might have occurred during averaging
```
Comment (mryab) [inline hivemind/client/averaging/training.py:86]:
```suggestion
                        # since tensors might have changed, we subtract old_local_tensor and add averaged. This prevents
                        # losing local updates that might have occurred during averaging
```

File: hivemind/client/averaging/training.py [hunk 2]
```diff
@@ -46,27 +48,51 @@ def __init__(self, opt: torch.optim.Optimizer, *, average_parameters: bool, aver
...
-            # load averaged tensors back into model
-            with self.get_tensors() as averaged_tensors:
-                assert len(averaged_tensors) == len(local_tensors)
-                for averaged_tensor, local_tensor in zip(averaged_tensors, local_tensors):
-                    local_tensor[...] = averaged_tensor.to(dtype=local_tensor.dtype, device=local_tensor.device)
+            if gathered is not None:
+                # load averaged tensors back into model
+                with data_lock, self.get_tensors() as averaged_tensors:
+                    assert len(averaged_tensors) == len(
+                        local_tensors), "The number of optimized parameters should not change."
```
Comment (mryab) [inline hivemind/client/averaging/training.py:82]: We'd better raise an error here: assertions are okay if they verify the function usage from a user POV, but in this case something went wrong inside the averaging loop

File: hivemind/optim/adaptive.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    Behaves exactly as CollaborativeOptimizer except:
+     - averages adaptive learning rates of an optimizer
+     - doesn't average gradients
+    :param average_opt_statistics: average optimizer statistics with corresponding names in statedict
+    :param kwargs: options for CollaborativeOptimizer
+    """
+
+    def __init__(self, average_opt_statistics: Sequence[str], **kwargs):
+        self.average_opt_statistics = average_opt_statistics
+        super().__init__(self, **kwargs)
```
Comment (mryab) [inline hivemind/optim/adaptive.py:18]: Init is usually called before setting the child attributes

File: hivemind/optim/adaptive.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from typing import Optional, Sequence
```
Comment (mryab) [inline hivemind/optim/adaptive.py:1]: Is Optional required?

File: hivemind/optim/simple.py [hunk 5]
```diff
@@ -1,55 +1,67 @@
...
+    A simple optimizer that trains a shared model by averaging with peers in variety of ways. Supports
+    parameter/gradient averaging and syncing adaptive learning rates or any other internal statistics of optimizer.
 
     :param opt: a pytorch optimizer configured to update model parameters.
     :param dht: a running hivemind DHT daemon connected to other peers
-    :param averaging_period: if specified, optimizer will attempt to average weights at regular intervals of this many
-      seconds. (averaging step will only occur if the optimizer ran at least one step in that interval)
+    :param average_parameters: whether to average model parameters
+    :param average_gradients: whether to average gradients
+    :param average_opt_statistics: if specified, average optimizer statistics with corresponding names in statedict
```
Comment (mryab) [inline hivemind/optim/simple.py:24]:
```suggestion
    :param average_opt_statistics: if specified, average optimizer states with corresponding names in state_dict
```

File: hivemind/optim/simple.py [hunk 6]
```diff
@@ -1,55 +1,67 @@
...
     :param prefix: all DHT keys that point to optimization metadata will have this prefix
     :param target_group_size: maximum group size for averaging (see DecentralizedAverager)
     :param timeout: if DecentralizedAverager step is unable to form group in this many seconds, cancel step
-    :param kwargs: additional parameters passed to DecentralizedAverager
-
-    :note: using DecentralizedOptimizer with adaptive learning rates may result in poor convergence due to
-      out-of-sync adaptive learning rates (such as adam second momentum or schedule step). Please ensure that these
-      statistics are synchronized or use a more advanced DecentralizedOptimizer version, if applicable.
+    :param kwargs: additional parameters passed to TrainingAverager
+    :note: if you're using optimizer with adaptive learning rates (such as adam second momentum), make sure to specify
```
Comment (mryab) [inline hivemind/optim/simple.py:32]:
```suggestion
    :note: if you're using an optimizer with adaptive learning rates (such as Adam), make sure to specify
```

**Review by mryab — CHANGES_REQUESTED** [review]
1. Check that all imports are actually used
2. Better to add a test for averaging optimizer states

File: hivemind/optim/adaptive.py (hunk 3 above)
Comment (nevec) [reply hivemind/optim/adaptive.py]: Fixed. Solution -  pass as parameter

**Review by justheuristic — APPROVED** [review]

File: tests/test_training.py [hunk 7]
```diff
@@ -135,3 +135,28 @@ def test_decentralized_optimizer_step():
...
+                            prefix='foo', target_group_size=2, verbose=True)
+
+    assert not torch.allclose(param1, param2)
+
+    (param1.sum() + param2.sum()).backward()
+
+    opt1.step()
+    opt2.step()
+
+    time.sleep(0.5)
```
Comment (mryab) [inline tests/test_training.py:160]: Do we really need to wait that long for the averaging to finish? Also, not for this PR, but it'd be great to have a synchronous version of averaging which blocks until finished or until the averaging fails

**Review by mryab — APPROVED** [review]
